File: packages/core/src/routes/organizations.ts
/**
 * Organization routes — thin HTTP layer over OrganizationService.
 */

import { Hono } from 'hono'
import type { Services } from '../services'
import { mapStatusCode } from './server'

export function organizationRoutes(services: Services) {
  const app = new Hono()

  app.post('/organizations', async (c) => {
    const body = await c.req.json()
    const result = services.organizations.create(body)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data }, 201)
  })

  app.get('/organizations', (c) => {
    const offset = Math.max(0, parseInt(c.req.query('offset') ?? '0', 10) || 0)
    const limit = Math.min(200, Math.max(1, parseInt(c.req.query('limit') ?? '50', 10) || 50))
    const filter: Record<string, string | number> = {}
    if (c.req.query('org_type')) filter.org_type = c.req.query('org_type')!
    if (c.req.query('tag')) filter.tag = c.req.query('tag')!
    if (c.req.query('worked') !== undefined && c.req.query('worked') !== null) {
      filter.worked = parseInt(c.req.query('worked')!, 10)
    }
    if (c.req.query('search')) filter.search = c.req.query('search')!
    if (c.req.query('status')) filter.status = c.req.query('status')!

    const result = services.organizations.list(filter, offset, limit)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data, pagination: result.pagination })
  })

  app.get('/organizations/:id', (c) => {
    const result = services.organizations.get(c.req.param('id'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.patch('/organizations/:id', async (c) => {
    const body = await c.req.json()
    const result = services.organizations.update(c.req.param('id'), body)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.delete('/organizations/:id', (c) => {
    const result = services.organizations.delete(c.req.param('id'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.body(null, 204)
  })

  return app
}

File: crates/forge-server/src/routes/organizations.rs
//! Organization CRUD routes.
//!
//! Mirrors the TS organization routes.

use axum::extract::{Path, Query, State};
use axum::routing::{get, post};
use axum::{Json, Router};
use serde::Deserialize;

use forge_core::{CreateOrganizationInput, Organization, OrganizationFilter};
use forge_sdk::db::OrganizationStore;

use crate::db::with_conn;
use crate::error::ApiError;
use crate::response::{ApiData, ApiList, Created, NoContent};
use crate::state::SharedState;

// ── Query params ────────────────────────────────────────────────────

#[derive(Debug, Deserialize, Default)]
pub struct OrgListQuery {
    pub offset: Option<i64>,
    pub limit: Option<i64>,
    pub org_type: Option<String>,
    pub tag: Option<String>,
    pub worked: Option<i32>,
    pub search: Option<String>,
    pub status: Option<String>,
}

// ── Handlers ────────────────────────────────────────────────────────

async fn create_organization(
    State(state): State<SharedState>,
    Json(input): Json<CreateOrganizationInput>,
) -> Result<Created<Organization>, ApiError> {
    let result =
        with_conn(&state, move |conn| OrganizationStore::create(conn, &input)).await?;
    Ok(Created(result))
}

async fn list_organizations(
    State(state): State<SharedState>,
    Query(q): Query<OrgListQuery>,
) -> Result<Json<ApiList<Organization>>, ApiError> {
    let filter = OrganizationFilter {
        org_type: q.org_type,
        tag: q.tag,
        worked: q.worked,
        search: q.search,
        status: q.status,
    };
    let offset = Some(q.offset.unwrap_or(0).max(0));
    let limit = Some(q.limit.unwrap_or(50).clamp(1, 200));

    let (data, pagination) = with_conn(&state, move |conn| {
        OrganizationStore::list(conn, Some(&filter), offset, limit)
    })
    .await?;

    Ok(Json(ApiList { data, pagination }))
}

async fn get_organization(
    State(state): State<SharedState>,
    Path(id): Path<String>,
) -> Result<Json<ApiData<Organization>>, ApiError> {
    let result = with_conn(&state, move |conn| {
        OrganizationStore::get(conn, &id)?.ok_or_else(|| forge_core::ForgeError::NotFound {
            entity_type: "Organization".into(),
            id: id.clone(),
        })
    })
    .await?;
    Ok(Json(ApiData { data: result }))
}

async fn update_organization(
    State(state): State<SharedState>,
    Path(id): Path<String>,
    Json(input): Json<CreateOrganizationInput>,
) -> Result<Json<ApiData<Organization>>, ApiError> {
    let result =
        with_conn(&state, move |conn| OrganizationStore::update(conn, &id, &input)).await?;
    Ok(Json(ApiData { data: result }))
}

async fn delete_organization(
    State(state): State<SharedState>,
    Path(id): Path<String>,
) -> Result<NoContent, ApiError> {
    with_conn(&state, move |conn| OrganizationStore::delete(conn, &id)).await?;
    Ok(NoContent)
}

// ── Router ──────────────────────────────────────────────────────────

pub fn router() -> Router<SharedState> {
    Router::new()
        .route(
            "/organizations",
            post(create_organization).get(list_organizations),
        )
        .route(
            "/organizations/{id}",
            get(get_organization)
                .patch(update_organization)
                .delete(delete_organization),
        )
}
